### PR TITLE
Standard loading indicator flow.

### LINKED ISSUE #598: Extend amp-img's "loading" functionality to support other AMP elements.

### PR DESCRIPTION
**Pull request by dvoytenko**
Closes #598.

### CODE REVIEW

File: spec/amp-html-layout.md [hunk 1]
```diff
@@ -112,6 +112,14 @@ is up to the element's implementation.
     </amp-anim>
 ```
 
+### `noloading`
```
Comment (erwinmombay) [inline spec/amp-html-layout.md]: :+1:

File: src/custom-element.js [hunk 2]
```diff
@@ -715,6 +790,95 @@ export function createAmpElementProto(win, name, implementationClass) {
...
+   */
+  ElementProto.isLoadingEnabled_ = function() {
+    // No loading indicator will be shown if either one of these
+    // conditions true:
+    // 1. `noloading` attribute is specified;
+    // 2. The element has not been whitelisted;
+    // 3. The element is too small or has not yet been measured;
+    // 4. The element has already been laid out;
+    // 5. The element is a `placeholder` or a `fallback`;
+    // 6. The element's layout is not a size-defining layout.
```
Comment (erwinmombay) [inline src/custom-element.js]: thanks for the clear explanation :100:

File: css/amp.css [hunk 3]
```diff
@@ -243,6 +249,28 @@ amp-video video {
   animation-delay: .2s;
 }
 
+@keyframes -amp-loader-dots {
```
Comment (erwinmombay) [inline css/amp.css]: i thought we were switching to spinner? just deferring?

Comment (dvoytenko) [inline css/amp.css]: Yes. This PR is pretty big.

File: src/custom-element.js [hunk 4]
```diff
@@ -715,6 +790,95 @@ export function createAmpElementProto(win, name, implementationClass) {
...
+      if (!this.loadingContainer_) {
+        return;
+      }
+
+      this.loadingContainer_.classList.toggle('amp-hidden', !state);
+      this.loadingElement_.classList.toggle('amp-active', state);
+
+      if (!state && opt_cleanup) {
+        let loadingContainer = this.loadingContainer_;
+        this.loadingContainer_ = undefined;
```
Comment (erwinmombay) [inline src/custom-element.js]: a little unsure as far as best practice, but is this more common than to `null` something out?

File: src/custom-element.js [hunk 5]
```diff
@@ -715,6 +790,95 @@ export function createAmpElementProto(win, name, implementationClass) {
...
+  /**
+   * Creates a loading object. The caller must ensure that loading can
+   * actually be shown. This method must also be called in the mutate
+   * context.
+   * @private
+   */
+  ElementProto.prepareLoading_ = function() {
+    if (!this.loadingContainer_) {
+      let container = document.createElement('div');
+      container.setAttribute('loading', '');
```
Comment (erwinmombay) [inline src/custom-element.js]: is the `loading` attribute used for anything?

File: src/custom-element.js (hunk 4 above)
Comment (dvoytenko) [inline src/custom-element.js]: Yeah. Ok. Switched to `null`. Seemed like `undefined` was warranted, but I agree.

File: src/custom-element.js (hunk 5 above)
Comment (dvoytenko) [inline src/custom-element.js]: Removed. Copy-paste-rename-forgot-it-was-there.

**Comment by erwinmombay** [comment]
@dvoytenko looks great! LGTM

**Pull request merged**